#!/usr/bin/env Rscript



# install.packages("systemfonts") # if you have not installed "devtools" package
install.packages("devtools") # if you have not installed "devtools" package
library(devtools)
devtools::install_github("BioinformaticsFMRP/TCGAbiolinks")
install.packages("pacman")
pacman::p_load(BiocManager, preprocessCore, DT, TCGAbiolinks, SummarizedExperiment, sesameData, sesame)

#### Codex #####################################################################
barcode_codex = list(
  "TCGA-LGG" = c("TCGA-CS-6665", "TCGA-DB-5270", "TCGA-FG-8187", "TCGA-HT-A619")
)
query_codex = list(
  "mirna" = list( category ="Transcriptome Profiling",data_type="miRNA Expression Quantification"),
  "mirna_isoform" = list( category ="Transcriptome Profiling",data_type="Isoform Expression Quantification"),
  "mrna" = list(category ="Transcriptome Profiling",data_type="Gene Expression Quantification"),
  "mrna_differential" = list(category ="Transcriptome Profiling",data_type="Differential Gene Expression"),
  "methylation" = list(category ="DNA Methylation",data_type="Methylation Beta Value"),
  "protein" = list(category ="Proteome Profiling",data_type="Protein Expression Quantification"),
  "cnv" = list(category ="Copy Number Variation",data_type="Gene Level Copy Number"),
  "cnv_segment" = list(category ="Copy Number Variation",data_type="Copy Number Segment"),
  "cnv_allele" = list(category ="Copy Number Variation",data_type="Allele-specific Copy Number Segment"),
  "snv" = list(category ="Simple Nucleotide Variation",data_type="Masked Somatic Mutation")
)
################################################################################


##### Variables to set #########################################################

cells_per_mb = 260412
temp_divider = 20
max_file_size = 100
#project <- "TCGA-LGG" #TCGA project from where to extract
#target <- "methylation"      #type of data to extract
#whole_data <- FALSE    #whether to use the whole dataset or just a few samples
#divider <- 1

input_message = "Input in format: Rscript file.R {project} {target} {use whole dataset} {divider} {path to data folder}
{path to layer file} {path to ids file}"

args = commandArgs(trailingOnly=TRUE) # get arguments from command line

if (length(args)!=7) {
  stop(input_message, call.=FALSE)
} else{
  project <- args[1]
  target <- tolower(args[2])
  whole_data <- as.logical(args[3])
  divider <- as.integer(args[4])
  data_folder <- args[5]
  layer_file <- args[6]
  ids_file <- args[7]
}

if (!(target %in% names(query_codex))){     #if target not in available options
  valid_options = paste(names(query_codex), collapse=", ")
  stop(paste("Invalid target\nValid options:", valid_options,"\n",input_message, sep=""), call.=FALSE)
}else if(is.na(whole_data)){
  stop(paste("Invalid whole_data flag\nValid options: TRUE, FALSE\n", input_message, sep=""), call.=FALSE)
}else if(is.na(divider)){
  stop(paste("Invalid divider\nValid options: Integers\n", input_message, sep=""), call.=FALSE)
}

query_param <- query_codex[[target]]
test_barcodes <- barcode_codex[[project]]

##### Remove files if new generation ###########################################
print("## Removing files ##")
# f <- paste(layer_folder, '/', target, "*", sep = "")
f <- gsub('.csv', '*', layer_file)
unlink(f)

##### Divider calculation ######################################################
if (divider == -1){
  print("## Calculating divider ##")
  ids <- readRDS(ids_file)    #gets ids in disk generated by get_ids.R
  barcodes <- split(ids, rep_len(1:temp_divider, length(ids))) #splits vector of ids by 20
  query <- TCGAbiolinks::GDCquery(project = project, data.category = query_param$category, 
                                  data.type = query_param$data_type, legacy = FALSE,
                                  barcode = unlist(barcodes[1]))
  

  TCGAbiolinks::GDCdownload(query = query, directory = data_folder, method = "api", files.per.chunk = 1)
  
  data <- TCGAbiolinks::GDCprepare(query = query, save = FALSE, directory = data_folder,
                                   summarizedExperiment = FALSE, remove.files.prepared = FALSE)

  n_cells <- nrow(data) * ncol(data) * temp_divider
  size_mb <- n_cells/cells_per_mb
  divider <- ceiling(size_mb/max_file_size)
  print(divider)
}

##### Dataset creation ###########################################################
if (whole_data){
  ids <- readRDS(ids_file)    #gets ids in memory generated by get_ids.R
  barcodes <- split(ids, rep_len(1:divider, length(ids))) #splits vector of ids by divider
}else{
  barcodes <- list(test_barcodes)
}

n_segments <- length(barcodes)

for (i in 1:n_segments){
  print(paste("\n#### Iteration", i,"####"))
  
  ##### Query ##################################################################
  print("## Creating query ##")
  query <- TCGAbiolinks::GDCquery(project = project, data.category = query_param$category, 
                                  data.type = query_param$data_type, 
                                  legacy = FALSE,
                                  barcode = unlist(barcodes[i]))
  
  ##### Download ###############################################################
  print("## Downloading files ##")
  TCGAbiolinks::GDCdownload(query = query, directory = data_folder, method = "api", files.per.chunk = 1)
  
  ##### Transform ##############################################################
  print("## Transforming data ##")
  data <- TCGAbiolinks::GDCprepare(query = query, save = FALSE, save.filename = paste(data_folder, "/temp.RData", sep=""),
                                   directory = data_folder, summarizedExperiment = FALSE, remove.files.prepared = FALSE)
  
  ##### Save ###################################################################
  print("## Saving data ##")
  segment_file <- gsub('.csv', paste('_', i, '.csv', sep = ''), layer_file)
  print(segment_file)
  if (is.data.frame(data)){
    # write.csv(data, paste(layer_folder, '/', target, "_", i, ".csv", sep=""), row.names = FALSE)
    write.csv(data, segment_file, row.names = FALSE)
  }else if (is.matrix(data)){
    data <- as.data.frame(data)
    data <- tibble::rownames_to_column(data, "ID")
    # write.csv(data, paste(layer_folder, '/', target, "_", i, ".csv", sep=""), row.names = FALSE)
    write.csv(data, segment_file, row.names = FALSE)
  }else if (class(data) == "RangedSummarizedExperiment" | class(data) == "SummarizedExperiment"){
    # write.csv(assay(data),paste(layer_folder, '/', target, "_", i, ".csv", sep=""), row.names = FALSE)
    # write.csv(rowData(data),paste(layer_folder,"/rowData_", target, "_", i, ".csv", sep=""), row.names = FALSE)
    write.csv(assay(data), segment_file, row.names = FALSE)
    write.csv(rowData(data), segment_file, row.names = FALSE)
  }
  
  rm(data)
}
  
#rm(list=ls())  #remove all variables
rm(list=setdiff(ls(), c("target","n_segments","layer_file")))  #remove all variable except

# ##### Join multiple datasets ###################################################
# print("## Joining multiple datasets ##")
# #target="methylation"
# #n_segments = 20
# #layer_folder <- "~/Desktop/thesis/data/TCGA-LGG/"
# 
# final <- read.csv(paste(layer_folder, target, "_1", ".csv", sep = ""), check.names = FALSE)
# id_col <- colnames(final)[1]
# 
# if (n_segments > 1) {
#   for (i in 2:n_segments){
#     print(paste("Joining segment:", i))
#     print(sort( sapply(ls(),function(x){object.size(get(x))})))
#     t <- read.csv(paste(layer_folder, target, "_", i, ".csv", sep = ""), check.names = FALSE)
#     final <- merge(final, t, by=id_col)
#     rm(t)
#   }
# }
# 
# 
# final <- final[rowSums(is.na(final)) != ncol(final),] #remove rows with all NaN
# write.csv(final, paste(layer_folder, target, ".csv", sep=""), row.names = FALSE)   #write final dataframe

rm(list=ls())  #remove all variables

############################################################################################################################################333

#pacman::p_unload(all)
#detach("package:", unload = TRUE)
#cat("\014")


